### ISSUE TITLE
Enhancement: [strict-boolean-expressions] option to ban strict equality on nullable boolean

### ISSUE THREAD

**Issue by scarf005** [typescript-eslint/typescript-eslint#9701]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/strict-boolean-expressions#allownullableboolean

### Description

as undefined | null is falsy, it's often desired to treat falsy values as false. however, it's easy to accidently strict comparing nullable boolean, which may not be desirable.

add a new option to `allowNullableBoolean`, `noStrictComparison`.

(context: `bool?: boolean`)

| `allowNullableBoolean` | effect | fail example | pass example |
| :-: | - | - | - |
| false | will error when comparing truthy/falsy nullable boolean | `if (bool) {}` | `if (bool === true() {}` |
| true | will accept when comparing truthy/falsy nullable boolean | - | `if (bool) {}` |
| noStrictComparison | will error when **strict** comparing falsy nullable boolean | `if (bool === false) {}` | `if (!bool) {}` |

### Fail

```TypeScript
const foo = (bool?: boolean) => {
  if (bool !== false) {}
  if (bool === false) {}
}
```

### Pass

```TypeScript
const foo = (bool?: boolean) => {
  if (bool) {}
  if (!bool) {}
}
```

### Additional Info

- https://typescript-eslint.io/rules/no-unnecessary-boolean-literal-compare/#allowcomparingnullablebooleanstofalse wouldn't work because the autofix converts the check to `bool ?? true`, which is opposite of what the issue suggests.

**Comment by bradzacher** [comment]
The rule purposely does not check overly coarse logicals. There's no way for the rule to know that you wanted `bool === false` to be `bool === false || bool == null`.

The point of the rule is to ensure you are handling a single truthy/falsey value per logical check - so expanding the rule's purview to enforce that you purposely always check multiple falsey states in a logical will not happen. I.e. at the surface this request is rejected.

---

> `no-unnecessary-boolean-literal-compare `'s `allowComparingNullableBooleansToFalse` wouldn't work because the autofix converts the check to `bool ?? true`, which is opposite of what the issue suggests

The reason it works that way is on purpose - like `strict-boolean-expressions` it's assuming you wrote the code like that on purpose and it's just fixing to remove the `=== false` comparison.

---

Personally I don't think this passes the non-niche check.
There is a lot of hidden complexity in your proposal - it's not possible for us to really tell what the intent was -- did the author intend to skip the nullish check or not? This would lead to the potential for a lot of false positives which can be quite frustrating.

cc @typescript-eslint/triage-team

**Comment by kirkwaiblinger** [comment]
Hmm, this has some overlap with [no-unnecessary-boolean-literal-compare](https://typescript-eslint.io/rules/no-unnecessary-boolean-literal-compare/). I wonder if that rule would get you what you're looking for, @scarf005 ?

*** 

This reminds me of another idea I have been meaning to propose, though, which is that I would like to be able to ban any (in)equality comparison in which either side is possibly nullish, unless it is to check for nullishness itself (i.e. `x == null`). I think this proposal would also be a subset of that, idea, possibly?

**Comment by kirkwaiblinger** [comment]
Oh, sorry, I missed the part where you already mentioned no-unnecessary-boolean-literal-compare.

**Comment by kirkwaiblinger** [comment]
i feel like you could still use SBE in conjunction with NUBLC... you'd just have to clean up after the autofix. Maybe with no-restrict-syntax? 


(ok, it ain't pretty)
```json
    "no-restricted-syntax": [
      "error", 
      {
        "selector": "LogicalExpression[operator=\"??\"][right.type=\"Literal\"][right.raw=\"true\"]",
        "message": "Ban `?? true`"
      },
      {
        "selector": "LogicalExpression[operator=\"??\"][right.type=\"Literal\"][right.raw=\"false\"]",
        "message": "Ban `?? false`"
      }
    ],
```

But give it a spin [in the playground](https://typescript-eslint.io/play/#ts=5.5.2&showAST=es&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3mpBbEAqiqAGYCWKIwAwqsGRmagFzwBGSSEIUK8APvACuxEOUrAA3ACgyJeAAoc%2BIqQpVaxBkz4BeffBJQIaEAEp4Ab2kBfadNCRYCZOizKQAOSEQIm%2Bows7JzcvALwKD4QMnKKHt6%2B-tqo8ACEBhgwQuZWtkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1xsciaNHzRIAGjD8BWAdkjJESGqO6QAMvgDm1NvACiHAA4lkyCvlw98RoW1KiAvAB1IAfjeuAuj0paAFqQAdKRMti4aFLTQet6%2BFAHBMQDuEeSwiN4S0nKQALYoyGxaiGoAQmy4YAAGHmDpiNWQORAAvuItmJ04isoOYhga2roGxqbmlta2Mf0RHnF%2BgSFhiBHqUXbwCwlLKREAZnqKWR1y8gVmxaWDFVW1bmCH8IpNna05XqfyAAKhtshUShGUiEFDwCisAD0BEIsFwuEQVEKbGgTEIACN8PgkJVCODonpCFR8HkjCjrrxOpAhCIxF8ZN0oHp4PhkgBhElkyi4LQAOQQ8DY6KQZSxONwyAAKvgAGJHClPRRvD78d6tIA&tsconfig=&tokens=false)?

**Comment by JoshuaKGoldberg** [comment]
> Personally I don't think this passes the non-niche check.

Agreed. Given that this hasn't received any positive interactions in >3 months, and there's a `no-restricted-syntax` workaround, and none of the team's comments are particularly positive, this is a wontfix for now.

Thanks for filing!
